it/src/test/java/com/google/cloud/teleport/it/spanner/SpannerResourceManagerUtilsTest.java

```diff
@@ -0,0 +1,166 @@
+ * the License.
+ */
+package com.google.cloud.teleport.it.spanner;
+
+import static com.google.cloud.teleport.it.spanner.SpannerResourceManagerUtils.generateDatabaseId;
+import static com.google.cloud.teleport.it.spanner.SpannerResourceManagerUtils.generateInstanceId;
+import static com.google.cloud.teleport.it.spanner.SpannerResourceManagerUtils.generateNewId;
+import static com.google.common.truth.Truth.assertThat;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Unit tests for {@link com.google.cloud.teleport.it.spanner.SpannerResourceManagerUtils}. */
+@RunWith(JUnit4.class)
+public final class SpannerResourceManagerUtilsTest {
+
+  @Test
+  public void testGenerateDatabaseId_replacesDigitLastCharWithLetter() {
+    String testBaseString = "test_database_0";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).matches("test_database_[a-z]");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_replacesDollarSignWithUnderscore() {
+    String testBaseString = "test$database";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_replacesDotWithUnderscore() {
+    String testBaseString = "test.database";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_replacesHyphenWithUnderscore() {
+    String testBaseString = "test-database";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_replacesNonLetterFirstCharWithLetter() {
+    String testBaseString = "0_test_database";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).matches("[a-z]_test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_replacesUpperCaseLettersWithLowerCase() {
+    String testBaseString = "Test_Database";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_trimsTrailingHyphen() {
+    String testBaseString = "test_database---";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateDatabaseId_trimsTrailingUnderscore() {
+    String testBaseString = "test_database___";
+
+    String actual = generateDatabaseId(testBaseString);
+
+    assertThat(actual).isEqualTo("test_database");
+  }
+
+  @Test
+  public void testGenerateInstanceId_replacesDollarSignWithHyphen() {
+    String testBaseString = "test$instance";
+
+    String actual = generateInstanceId(testBaseString);
+
+    assertThat(actual).matches("test-instance-[0-9]{8}-[0-9]{6}-[0-9]{6}");
```

Comment:
nit: `\d` is a cleaner way of matching any number than `[0-9]`. Same goes for other uses.